src/cpp/rtps/reader/StatefulReader.cpp

```diff
@@ -612,6 +619,7 @@ bool StatefulReader::processHeartbeatMsg(
                     if (liveliness_kind_ == MANUAL_BY_TOPIC_LIVELINESS_QOS ||
                             writer->liveliness_kind() == MANUAL_BY_TOPIC_LIVELINESS_QOS)
                     {
+                        lock.unlock(); // Avoid deadlock with LivelinessManager.
```

Comment:
If the deadlock is with the liveliness manager `assert_liveliness` method, I would move this line inside the `if (wlp != nullptr)`

If the deadlock is with the `wlp()` getter, it should stay here.